nds-org/workbench-helm-chart
# Deploy Workbench React UI + Python API + Keycloak (and friends) via Helm chart

## Problem
Workbench V2 will is ready for testing, but needs to be deployed via Helm chart

## Approach
Start adjusting the Helm chart with the changes necessary to deploy the new React UI for Workbench

## How to Test
Prerequisites:
* `helm` v3 + `kubectl` installed, with kubeconfig pointing to a running Kubernetes cluster
* `docker` (optional), `git`, `make` installed
* a fresh incognito browser window

See docs in README.md regarding Advanced Configuration to import a `realm.json` into Keycloak
See docs in README.md regarding Advanced Configuration to enable real TLS using `cert-manager` (selfsigned works for local development)

### Helm + Kube workflows (cluster admin)
1. Checkout this branch locally: `git clone https://github.com/nds-org/workbench-helm-chart; cd workbench-helm-chart/ && git fetch --all && git checkout v2-first-pass`
2. (optional) Adjust Makefile config parameters in the script (e.g. `NAME`, `NAMESPACE`, etc): `vi Makefile`
    * By default, `NAME=workbench` and `NAMESPACE=workbench`. This will install a Helm release named `workbench` in the `workbench` namespace of the Kubernetes cluster
3. (optional) Adjust `values.yaml` to fit your existing cluster's setup (e.g. `config.backend.domain`, `config.backend.keycloak`, etc)
4. Run `make` - this will run with sane defaults that may or may not work in a local environment
    * You should see the builder run `make dep` to pull dependency subcharts
    * You should see the builder run `make install` to install the Helm release on the cluster
5. Run `make local`
    * You should see that the platform is redeployed to use the full local dev setup, including a built-in Keycloak instance and imported realm (`make realm_import_secret` should be automatically run as well, in this case)
5. Run `make status`
    * You should see the status of Pods in the `workbench` namespace
    * Wait for all Pods to start
7. Navigate directly to `kubernetes.docker.internal` in your browser
    * You will see a warning about unsafe TLS certs (NOTE: this can be safely ignored on a `localhost` setup)
    * After proceeding past the warning, you should see the Workbench React UI
    * You should see no `_oauth2_proxy` cookie present (user is not logged in)
8. Navigate directly to `kubernetes.docker.internal/oauth2/userinfo` in your browser
    * You should see an `Unauthorized` message (user is not logged in)
9. Click "Log In"
    * You should be redirected to Keycloak to login
    * You should be offered the option to "Register" a new user
10. Click the "Register" link and signup a new user
    * You should be redirected to the Workbench React U
    * You should see that an `_oauth2_proxy` cookie is now present
11. Navigate once again to `kubernetes.docker.internal/oauth2/userinfo` in your browser
    * You should see now see a set of decoded user fields:
```json
{"user":"","email":"test@localhost","groups":["role:default-roles-workbench-dev","role:offline_access","role:uma_authorization"],"preferredUsername":"test"}
```
12. Run `make uninstall` to shut down Workbench, and wait for all pods to disappear (hint: `make watch` will watch for Pod events)
13. Run `make clean` to remove all PVCs in the `workbench` namespace
14. Run `make clean_all` to remove the entire `workbench` namespace

### Git + Docker workflows (local developer)
1. Run `make clone`
    * You should see git clone two repos (if they don't already exist locally)
2. Run `make pull`
    * You should see git pull from upstream on both local repos
2. Run `make dev`
    * You should see that the running instance is replaced with a version mapped in from local source
    * You should see that your changes are auto-compiled and used in the live running instance of Workbench
4. Run `make build`
    * You should see the builder builds Docker images from both of these repos
5. Run `make push`
    * You should see the builder push these two local Docker images
6. Run `make restart`
    * You should see the builder push these two local Docker images

See docs in README.md regarding Advanced Configuration to map this source code into running containers

## TODOs
* source chart somewhere.. NCSA Harbor? github pages?
* Verbose configuration documentation - already needs an update
* ~~`wait-for` startup ordering for `keycloak` <- `oauth2-proxy` + `apiserver` (OIDC discovery)~~
* ~~`wait-for` startup ordering for dependencies~~
* ~~mount user home volumes, if enabled~~
* ~~mount shared data volumes, if enabled~~
* ~~Adjust webui/apiserver to speak `_oauth2_proxy` instead of / addition in speaking `keycloak` for OIDC~~
* ~~Git release workflows? CI? Github Actions?~~
